Release process ADR

de-code opened the pull request:
First draft of the suggested ADR, see [initial discussions](https://github.com/elifesciences/issues/issues/4575).

de-code committed:
added initial release process adr

de-code commented on adr/0010-release-process.md:
```diff
@@ -0,0 +1,25 @@
+
+## Context
+
+eLife has some projects that are published as packages, e.g. for Python projects on pypi.
```
Not sure if we should try to make this ADR language agnostic or whether it's better to stick to Python here. Java for example has a more defined release process (e.g. including SNAPSHOT versions).

de-code commented on adr/0010-release-process.md:
```diff
@@ -0,0 +1,25 @@
+
+## Decision
+
+Each releasable project will have a mechanism to report on its current version. A release should be pushed, if there isn't currently a tag for the corresponding version (and it's on the master branch). i.e. the version in the develop branch will contain the old version until a new release version should be pushed.
```
Keeping the old version until the release seems to be relatively simple. Although semantically it seems counter intuitive:
a) Checking the version of a particular commit would include a release version without indication that it's something between
b) The code may already have breaking changes which are not reflected in the version. To me it seems the major version should change when we make breaking changes, not at release time.

(But then I don't want to engineer a complicated release process either)

giorgiosironi replied:
It may be more complicated yes. If you keep the old version around on an unstable `master` a package generated from that is indistinguishable from the stable version. In the PHP world I've seen people avoiding the version into the code, and this also helps having more than one release from the same commit e.g. `7.2.0-rc6` of PHP becomes the first stable release `7.2.0`. Needs more examples from the open source world.

de-code replied:
I'd be happy not to include the version in code, although many examples seem to have it. If it's not in the repo, then I guess the release would be manually triggered? (which seem to make sense for releases) That could be done via a parameterised Jenkins task or potentially via the GitHUb release mechanism.

giorgiosironi replied:
Talked in person today, seems to agree on:
- do not include a `__version__` field in the code to avoid the aforementioned problems on multiple packages with the same number
- use `git tag` to indicate a new release, fairly standard in open source
- Jenkins or other systems like Travis CI are able to trigger on the tag in some way to not just run tests but also build a package, push an image, etc (implementation detail)

de-code commented on adr/0010-release-process.md:
```diff
@@ -0,0 +1,25 @@
+
+Specific details may vary from project to project. For example one project may just pushed to pypi, another to Docker Hub, or even both. Some artefacts may just be pushed for releases, others on commit.
+
+The Build Infrastructure will need to support the release process, including managing the credentials to push the artefact. For that the Build Infrastructure will also need to know when a release should be pushed.
```
Managing the credentials should maybe be a separate ADR?

lsh-0 commented:
I would avoid baking version information into the code if at all possible. Let the repository manage version information, tags, branches, etc. That makes the mechanism universal to all code.

Generation of artefacts would occur after a release has been tagged and pushed to whatever repositories (maven, pypi, github, etc).

The release process should be either handcrafted by a human or completely automated. I think we should be quite mercenary about this. If you decide to have a human intervene and decide release numbers, then there should be a small number of people doing release management - it shouldn't be a (another) responsibility pushed on to the developer. As always, I'm highly in favour of a completely automated process.

Do you see a distinction between the build process and the release process?

Regarding the ADR, I'd like to see:

* a description of what the desired release process would look like.
* arguments for a formal build release vs our current state. It may seem obvious to us but pretend you're writing for an audience that isn't technical. What do we gain from an x.y.z number vs a sha?
* keep implementation details out of the ADR, like where credentials are stored or names of specific programs (Alfred).

giorgiosironi commented:
> keep implementation details out of the ADR, like where credentials are stored or names of specific programs (Alfred).

Agreed on discussing this elsewhere

> Generation of artefacts would occur after a release has been tagged and pushed to whatever repositories (maven, pypi, github, etc).

In general the direction of information should be Git -> Jenkins (or any other build and release process) as Git is universal to operate (push a `git tag`, etc.) and remains around as an audit log.

Speaking about possible implementations, https://alfred.elifesciences.org/job/pull-requests-projects/configure has a `Discover tags` option where it can build tags in addition to pull requests. Need to experiment a bit with it but we should be able to detect when a tag is issued and then run parts of the build only in that case (as opposed to `develop` or PRs).

lsh-0 commented:
I'm fine with Alfred handling the release management stuff. I guess it's the process I want made crystal clear by the ADR. Flow diagrams might be handy. I wrote at length a little while ago (but elsewhere) on possible approaches to all this and I'd like to see what you guys are recommending so I can reconcile it with my own thoughts. I think it's a good idea overall. It's mostly for xpub and non-elifers, right?

de-code committed:
removed comment regarding infrastructure

de-code committed:
updated decision

de-code commented:
I removed the infrastructure part and updated the decision.

I'm not really sure whether I am the right person to "own" this ADR. Would anyone else want to volunteer?

giorgiosironi commented:
I'll take ownership of this.

de-code commented:
Just want to check whether there are any updates on this? Do we maybe have some other reference projects already that include releases?

lsh-0 commented:
> Do we maybe have some other reference projects already that include releases?

looks like Sean might have started doing them before he left:

* https://github.com/elifesciences/elife-tools/releases

Outside of elife (and non-python projects) I'm doing them manually. The `master` branch is always the latest release, the `develop` branch accumulates changes until an arbitrary point and then then a PR is opened, the changes merged in to `master` and `master` is tagged following semver guidelines. Travis picks up on the tag and does some extra steps after testing to generate a release and push it to Github. I'm considering automated nightly builds so changes are released more frequently if the build is passing.

lsh-0 commented:
Regarding baking version information into the code, I seem to be pretty against it in previous comments, but I guess if there is an elegant way to manage it that doesn't involve code and anything too-dynamic, it should be ok ...? The impulse to specify it in code (`version = "1.0.0"`) inevitably leads to other weird convolutions. 

I'm keen to know what you guys settle on, my own process is still evolving

giorgiosironi commented:
> Regarding baking version information into the code, I seem to be pretty against it in previous comments

I agree with this now (not sure if I did originally.) Revisiting this ADR today.

lsh-0 commented:
Yeah, it's always been troublesome for me. I have something that works for me in a very particular situation right now but that probably wouldn't translate to other languages/environments.

giorgiosironi commented on adr/0010-release-process.md:
```diff
@@ -0,0 +1,22 @@
+## Decision
+
+A new release will be indicated by a new git tag.
+Jenkins will trigger on a new tag to, beyond building and running tests, also push the release out (e.g. pypi, Docker Hub, etc.).
```
investigating this bit, which I have never used in anger

giorgiosironi replied:
Turned on build of tags not older than 7 days, at https://alfred.elifesciences.org/job/pull-requests-projects/computation/console

giorgiosironi replied:
Tag ignored by the scan, still located at https://alfred.elifesciences.org/job/pull-requests-projects/computation/console

giorgiosironi replied:
Also had to add `Discover tags` to the branch sources.

giorgiosironi replied:
Discovered, but didn't build until I manually triggered it: https://alfred.elifesciences.org/job/pull-requests-projects/job/elife-jenkins-workflow-libs/view/tags/job/0.0.1/

giorgiosironi replied:
Seems these jobs have a `TAG_NAME=0.0.1` environment variable we can use to check.

giorgiosironi replied:
https://github.com/elifesciences/elife-jenkins-workflow-libs/pull/14

giorgiosironi replied:
https://alfred.elifesciences.org/job/pull-requests-projects/job/elife-jenkins-workflow-libs/view/tags/job/0.0.2/ correctly triggered after being pushed (the previous one was already existing so only discovered but not built).

giorgiosironi replied:
Also https://github.com/elifesciences/elife-jenkins-workflow-libs/pull/15

giorgiosironi commented on adr/0010-release-process.md:
```diff
@@ -0,0 +1,22 @@
+
+A new release will be indicated by a new git tag.
+Jenkins will trigger on a new tag to, beyond building and running tests, also push the release out (e.g. pypi, Docker Hub, etc.).
+
```
This doesn't prescribe how to deal with `master` and `latest` and whether a tag is always also pushed as something like that. Out of scope for this ADR, so I'll rename it to be more precise in scope. And add consequences.

Commit:
State importance of versioning

Commit:
Add consequences

Commit:
Change status to Proposed

Commit:
Rename to reflect scope

giorgiosironi commented:
Updated with lessons learnt from https://github.com/elifesciences/sciencebeam/pull/126

thewilkybarkid commented on adr/0010-release-via-tags.md:
```diff
@@ -0,0 +1,27 @@
+
+A new release will be indicated by a new git tag.
+
+Jenkins will trigger on a new tag to, beyond building and running tests, also push the release out (e.g. pypi, Docker Hub, etc.).
```
```suggestion
Jenkins will trigger on a new tag to, beyond building and running tests, also push the release out as needed (e.g. pypi, Docker Hub, etc.).
```

eg GitHub already notifies Packagist.

thewilkybarkid replied:
```suggestion
Jenkins will trigger on a new tag to, beyond building and running tests, also push the release out (e.g. pypi, Docker Hub).
```

thewilkybarkid commented on adr/0010-release-via-tags.md:
```diff
@@ -0,0 +1,27 @@
+
+## Decision
+
+A new release will be indicated by a new git tag.
```
```suggestion
A new release will be indicated by a new Git tag.
```

thewilkybarkid committed:
Update adr/0010-release-via-tags.md

Co-Authored-By: giorgiosironi <accounts@giorgiosironi.com>

thewilkybarkid committed:
Update adr/0010-release-via-tags.md

Co-Authored-By: giorgiosironi <accounts@giorgiosironi.com>

Commit:
Make release publishing optional

diversemix commented on adr/0010-release-via-tags.md:
```diff
@@ -0,0 +1,27 @@
+
+## Consequences
+
+Version numbers cannot be specified in source code, as the tag is the source of truth for them.
```
Perhaps not specified... but you could certainly refer to old version numbers to keep behaviour consistent with older versions, which I see as the main reason for using version numbers in source.

lsh-0 replied:
git tags can be deleted and reused. For example you could have the tag 'latest' that always points to the most recent passing build. They're not immutable like other aspects of git

giorgiosironi replied:
No, branches have this behavior; tags are supposed to be immutable and it's very complicated to [overwrite them](https://git-scm.com/docs/git-tag#_on_re_tagging) in clients that already have the tag.

I'll rephrase to make clear that the code can refer to version numbers in general, but shouldn't contain its own current version number.

lsh-0 replied:
this link is also helpful:
https://git-scm.com/book/en/v2/Git-Basics-Tagging

Commit:
Clarify we are talking about the current version
